modules/host/contractmanager/writeaheadlog.go

```diff
@@ -104,34 +129,137 @@ type (
+		log.Print(err)
+		return build.ExtendErr("error decoding WAL metadata", err)
+	}
+
+	// Check if correct header was found
 	if md.Header != walMetadata.Header {
+		log.Print(err)
 		return errors.New("WAL metadata header does not match header found in WAL file")
 	}
 	if md.Version != walMetadata.Version {
+		log.Print(err)
 		return errors.New("WAL metadata version does not match version found in WAL file")
 	}
 	return nil
 }
 
 // writeWALMetadata writes WAL metadata to the input file.
-func writeWALMetadata(f file) error {
+func writeWALMetadata(f file) (int64, error) {
 	changeBytes, err := json.MarshalIndent(walMetadata, "", "\t")
 	if err != nil {
-		return build.ExtendErr("could not marshal WAL metadata", err)
+		return 0, build.ExtendErr("could not marshal WAL metadata", err)
 	}
-	_, err = f.Write(changeBytes)
+
+	_, err = f.WriteAt(changeBytes, headerLength())
+	if err != nil {
+		return 0, build.ExtendErr("unable to write WAL metadata", err)
+	}
+	return int64(len(changeBytes)), nil
+}
+
+// headerLength is a helper function that returns the changeLength of the binary encoded wal header
+func headerLength() int64 {
+	return int64(binary.Size(walHeader{}))
+}
+
+func prefixLength() int64 {
+	return int64(binary.Size(stateChangePrefix{}))
+}
+
+// readWALHeader reads the WAL header from the input file
+func readWALHeader(f file, header *walHeader) error {
+	bytesHeader := make([]byte, headerLength())
+
+	_, err := f.ReadAt(bytesHeader, 0)
 	if err != nil {
-		return build.ExtendErr("unable to write WAL metadata", err)
+		return build.ExtendErr("could not read WAL header", err)
+	}
+	buffer := bytes.NewBuffer(bytesHeader)
+
+	err = binary.Read(buffer, binary.LittleEndian, header)
+	if err != nil {
+		return build.ExtendErr("could not decode WAL header", err)
+	}
+	return nil
+}
+
+// writeWALHeader writes the WAL header to the input file
+func writeWALHeader(f file, header walHeader) error {
+	buffer := new(bytes.Buffer)
+
+	err := binary.Write(buffer, binary.LittleEndian, &header)
+	if err != nil {
+		return build.ExtendErr("could not marshal WAL header", err)
+	}
+	_, err = f.WriteAt(buffer.Bytes(), 0)
+	if err != nil {
+		return build.ExtendErr("could not write WAL header", err)
 	}
 	return nil
 }
 
+// readStateChange reads a stateChange at a specific offset in the file and
+// returns the changeLength of the stateChange read
+func readStateChange(f file, offset int64, sc *stateChange) (int64, error) {
+	// Read the prefix
+	bytesPrefix := make([]byte, prefixLength())
+
+	_, err := f.ReadAt(bytesPrefix, offset)
+	if err != nil {
+		return 0, build.ExtendErr("could not read stateChange prefix", err)
+	}
+
+	// Decode the prefix
+	buffer := bytes.NewBuffer(bytesPrefix)
+	var prefix stateChangePrefix
+
+	err = binary.Read(buffer, binary.LittleEndian, &prefix)
+	if err != nil {
+		return 0, build.ExtendErr("could not decode stateChange prefix", err)
+	}
+
+	// Read the state change
+	bytesChange := make([]byte, prefix.ChangeLength)
```

Comment:
Just noticed that this is actually quite dangerous. Might have to check that `prefix.ChangeLength` is below `offset+wal.header.OffsetSC` to prevent a panic